### PR TITLE
Made DID Resolution language more relaxed

### LINKED ISSUE #186: DID you?
The current text [here in -01](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-01.html#section-3.5-2.2) is basically a normative requirement to support all DID resolution. Which is too much, even to the extent we want to allow for or support some DID usage. We need to revisit how this is phrased in the document and even how much (if any) is said. 

---------
created from discussion in PR #183 : 

B: "IMHO DIDs can be accounted/allowed for via the "Separate specifications or ecosystem regulations MAY define rules complementing the rules defined above" criteria below and shouldn't receive 1st order treatment like this in the spec."

O: "We added DIDs to make the spec useful also for the credential market that uses DIDs. For those people it is required to demonstrate how that can be done. I'm wondering how other folks are thinking about dropping DIDs from getting 1st order treatment. You are right, one can create another draft that just explains how to use DIDs with this spec."

_Originally posted by @awoie in https://github.com/oauth-wg/oauth-sd-jwt-vc/pull/183#discussion_r1387268502_

### PR DESCRIPTION
**Pull request by awoie**
This PR introduces the following changes:
- fixes #186 
- Made all public key validation rules conditional
- Rejecting the SD-JWT VC if public key validation cannot be done

### CODE REVIEW

**Comment by peacekeeper** [comment]
This harms interoperability, since some implementations will use DIDs and others will use `/.well-known/jwt-issuer`. A more interoperable approach would be to drop `/.well-known/jwt-issuer` and to always use DIDs. Those who want to use domain names and web servers can then still use `did:web`.

DIDs have always been designed as an abstraction layer for different types of resolvable identifiers. To build yet another layer of optionality, where DIDs are one option out of others, does not make much sense. Now you will have to deal with different metadata formats (JWT Issuer Metadata, and DID document).

Similar arguments apply to the `did:jwk` vs `cnf` discussion.

But I already know the response.. "Simple is a feature", and "DIDs have more boilerplate that is not needed", right? :)

**Comment by awoie** [comment]
> This harms interoperability, since some implementations will use DIDs and others will use `/.well-known/jwt-issuer`. A more interoperable approach would be to drop `/.well-known/jwt-issuer` and to always use DIDs. Those who want to use domain names and web servers can then still use `did:web`.
> 
> DIDs have always been designed as an abstraction layer for different types of resolvable identifiers. To build yet another layer of optionality, where DIDs are one option out of others, does not make much sense. Now you will have to deal with different metadata formats (JWT Issuer Metadata, and DID document).
> 
> Similar arguments apply to the `did:jwk` vs `cnf` discussion.
> 
> But I already know the response.. "Simple is a feature", and "DIDs have more boilerplate that is not needed", right? :)

@peacekeeper  Let's have the discussion in #186. As long as we haven't resolved the discussion I keep the DO NOT MERGE label on it.

**Comment by awoie** [comment]
> > This harms interoperability, since some implementations will use DIDs and others will use `/.well-known/jwt-issuer`. A more interoperable approach would be to drop `/.well-known/jwt-issuer` and to always use DIDs. Those who want to use domain names and web servers can then still use `did:web`.
> > DIDs have always been designed as an abstraction layer for different types of resolvable identifiers. To build yet another layer of optionality, where DIDs are one option out of others, does not make much sense. Now you will have to deal with different metadata formats (JWT Issuer Metadata, and DID document).
> > Similar arguments apply to the `did:jwk` vs `cnf` discussion.
> > But I already know the response.. "Simple is a feature", and "DIDs have more boilerplate that is not needed", right? :)
> 
> @peacekeeper Let's have the discussion in #186. As long as we haven't resolved the discussion I keep the DO NOT MERGE label on it.

Let me try to update the PR.

**Comment by awoie** [comment]
> > > This harms interoperability, since some implementations will use DIDs and others will use `/.well-known/jwt-issuer`. A more interoperable approach would be to drop `/.well-known/jwt-issuer` and to always use DIDs. Those who want to use domain names and web servers can then still use `did:web`.
> > > DIDs have always been designed as an abstraction layer for different types of resolvable identifiers. To build yet another layer of optionality, where DIDs are one option out of others, does not make much sense. Now you will have to deal with different metadata formats (JWT Issuer Metadata, and DID document).
> > > Similar arguments apply to the `did:jwk` vs `cnf` discussion.
> > > But I already know the response.. "Simple is a feature", and "DIDs have more boilerplate that is not needed", right? :)
> > 
> > 
> > @peacekeeper Let's have the discussion in #186. As long as we haven't resolved the discussion I keep the DO NOT MERGE label on it.
> 
> Let me try to update the PR.

@peacekeeper I made all rules now conditional from a verifier perspective. In this way a verifier is compliant if they don't support all rules but the SD-JWT VC would fail validation. Note, it is impossible for a verifier to support all rules that may be defined by ecosystems anyways. Are you ok with the updated language?

**Comment by peacekeeper** [comment]
> Are you ok with the updated language?

As a DID maximalist, I would personally prefer DIDs to be the only mechanism that is used whenever public keys are needed, instead of allowing more static public key mechanisms like "jwt-issuer" or "cnf".

Unlike "jwt-issuer", DIDs support identifiers that are persistent, decentralized, and cryptographically verifiable.

Unlike "cnf", DIDs support key rotation and other updates.

And as I said above, with DIDs there would be only one syntax and one resolution mechanism and one metadata format, instead of multiple different ones that individual implementations may or may not support.

But I also understand that for some communities, "keeping it simple" is more important than the above properties, so yes I'm fine with the language, thanks for the effort to update it!

**Comment by paulbastian** [comment]
While I understand your points, I think some of them are not true.

DIDs may be decentralized, may support key rotations and may have other features. There are enough examples of DID methods that do not have those properties. 

Making DIDs mandatory would limit optionality. But also with DIDs the optionality in my view shifts just shifts a layer further down as there are many DID methods. Indeed allowing DIDs expands optionality by 200+ did methods, so this could indeed be a counterargument.

"Cnf" isn't actually far away from DIDs, as the content can vary, there is a registry at IANA for supported methods, with different properties. In my view the cleanest approach for the keybinding would be to register a new cnf type "did" in this registry. But that's off topic in this PR.

I'm fine to keep some optionality in the specification at the beginning and then see what is actually being used

**Review by peppelinux — APPROVED** [review]

File: draft-ietf-oauth-sd-jwt-vc.md [hunk 1]
```diff
@@ -323,23 +323,22 @@ specification.
 
 ## Obtaining Public Key for Issuer-signed JWTs {#public-key-discovery-for-issuer-signed-jwts}
 
-A recipient of an SD-JWT VC MUST apply the following rules to obtain the public
-verification key for the Issuer-signed JWT:
+A recipient of an SD-JWT VC MUST apply the following rules to validate that the public
+verification key for the Issuer-signed JWT corresponds to the `iss` value:
 
-- JWT Issuer Metadata: If the `iss` value contains an HTTPS URI, the recipient MUST
+- JWT Issuer Metadata: If a recipient supports JWT Issuer Metadta and if the `iss` value contains an HTTPS URI, the recipient MUST
```
Comment (peppelinux) [inline draft-ietf-oauth-sd-jwt-vc.md:329]:
```suggestion
- JWT Issuer Metadata: If a recipient supports JWT Issuer Metadata and if the `iss` value contains an HTTPS URI, the recipient MUST
```

**Review by bc-pi — APPROVED** [review]
There are some things in this area that still need fixing which are done in #183 and is gonna have conflicts.  But assuming that can be done and resolved, I'm okay with the changes here.

File: draft-ietf-oauth-sd-jwt-vc.md [hunk 2]
```diff
@@ -678,6 +677,9 @@ for their contributions (some of which substantial) to this draft and to the ini
 
 # Document History
 
+-02
+* Made specific rules for public verification key validation conditional
```
Comment (bc-pi) [inline draft-ietf-oauth-sd-jwt-vc.md:681]:
```suggestion
-02

* Made specific rules for public verification key validation conditional
```

**Review by danielfett — APPROVED** [review]

**Pull request merged**